### ISSUE TITLE
Epoch number does not increase when training ann

### ISSUE THREAD

**Issue by shrit** [mlpack/mlpack#2073]
Hello, 

I training a forward neural network, two hidden layers and Adam optimizer. I put the train function inside a for loop to improve learning and reduce loss.
I am wondering if it is necessary to use this loop. In fact after adding the callbacks `PrintLoss()` `ProgressBar()` I have noticed that the epoch number does not increase during the training and still fixes on `1/29`.
The trained model is working well, but I am wondering whether the training procedure is correct or not:

```
 ens::AdamType<ens::AdamUpdate> optimizer;
 std::cout << "Parameters Optimizer: "<< model.Parameters() << std::endl;
  // Train the model.
  for (int i = 0; i < 300; ++i) {
    std::cout << "training..." << std::endl;
    model.Train(trainData,
                trainLabelsTemp,
                optimizer,
                ens::PrintLoss(),
                ens::ProgressBar(),
                ens::EarlyStopAtMinLoss(),
                ens::StoreBestCoordinates<arma::mat>());

    // optimizer.ResetPolicy() = false;

    arma::mat pred;
    model.Predict(trainData, pred);
    model.Evaluate(trainData, pred);
  }

```

**Comment by zoq** [comment]
Hello @shrit so I just tested this one on:

https://github.com/mlpack/mlpack/blob/c17b731731dcf75d1a314608edcc5068d56e1a6d/src/mlpack/tests/feedforward_network_test.cpp#L44

meaning I passed: ``ens::PrintLoss(), ens::ProgressBar(), ens::EarlyStopAtMinLoss(), ens::StoreBestCoordinates<arma::mat>()`` to the ``Train`` method and I can't see the situation you described above. So maybe this is specific to your model, what I think could happen is that ``EarlyStopAtMinLoss`` returns in the first epoch, can you remove ``ens::EarlyStopAtMinLoss()`` and test again?

Also, setting ``optimizer.ResetPolicy() = false;`` makes sense especially if you use an optimizer that update states during the optimization process like Adam.

**Comment by shrit** [comment]
@zoq  Thanks for your answers,
In fact, I have set already ` optimizer.ResetPolicy() = false;` but I commented it just before to test without it.
Here is the code of the model, just to make everything clear:
```

 // Initialize the network.
  FFN<MeanSquaredError<>, RandomInitialization> model;
  model.Add<Linear<> >(trainData.n_rows, 200);
  model.Add<LeakyReLU<> >();
  model.Add<Linear<> >(200, 200);
  model.Add<LeakyReLU<> >();
  model.Add<Dropout<> >(0.5);
  model.Add<Linear<> >(200, 200);
  model.Add<LeakyReLU<> >();
  model.Add<Dropout<> >(0.5);
  model.Add<Linear<> >(200, 3);

  ens::AdamType<ens::AdamUpdate> optimizer;

  std::cout << "Parameters Optimizer: "<< model.Parameters() << std::endl;

  // Train the model.
  for (int i = 0; i < 300; ++i) {
    // std::cout << "training..." << std::endl;
    model.Train(trainData,
                trainLabelsTemp,
                optimizer,
                ens::PrintLoss(),
                ens::ProgressBar(),
                ens::StoreBestCoordinates<arma::mat>());

    optimizer.ResetPolicy() = false;

    arma::mat pred;
    model.Predict(trainData, pred);
    model.Evaluate(trainData, pred);
  }
  arma::mat predtest;

  model.Predict(testData, predtest);
```

Actually the problem still here even when removing the `ens::EarlyStopAtMinLoss()` This is the output from the terminal when training:
```
Epoch 1/29
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 1115.75
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 3.17138
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 1.38296
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 1.196494
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 0.859519
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 0.521697
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 1.057025
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 0.190352
Epoch 1/29[=========================================================================================>..........] 89% - ETA: 0s - loss: 0.1114941

```
Looking at the above output why the Epoch number does not increase? I believe it has to reach at least 29/29 to train the model on the entire dataset.

**Comment by zoq** [comment]
An epoch is an iteration over the entire data, so ``29/29`` would mean you trained the model 20 x over the entire dataset. However a single epoch should reach ``100%``, is printed output above from a single iteration or are this multiple iterations (``i = 0, i=1, i=2, ... i=300``)?

Besids ``EarlyStopAtMinLoss`` the training will also abort if the minimal objectiv is reached:

https://github.com/mlpack/ensmallen/blob/7c37b614e35bd1e8f2131d5ee631013493be53bf/include/ensmallen_bits/sgd/sgd_impl.hpp#L167 so you might want to adjust the tolerance parameter for the optimizer as well:

```c++
optimizer.Tolerance() = 0.0000001;
```
or 

```c++
optimizer.Tolerance() = -1;
```

The model looks okay, not sure if you can provide the data as well?

**Comment by shrit** [comment]
In fact, the output lines are from the `for loop (i = 0, i=1, i=2, ... i=300)`
If I put `i < 1`, I will have only the first lines printed.
I have never seen `29/29`, only `1/29`. Also, it will never print `100%`. it only prints `89%.`
I have set `optimizer.Tolerance() = -1;`
I do not understand why the model.Train() return after only one epoch, can we change the number of epoch??
Also, the ETA is equal to 0 during the progress of the bar.

To make things clear, (when everything is fine), I do not have to use a for loop on the train function, suppose the epoch has reached 100/100. This will mean that the model has been trained 100 times over the dataset (Like Keras `fit` function)

**Comment by zoq** [comment]
I think what happens here is that:

https://github.com/mlpack/ensmallen/blob/7c37b614e35bd1e8f2131d5ee631013493be53bf/include/ensmallen_bits/sgd/sgd_impl.hpp#L176

is true, and the optimization process is stopped after the first epoch; that the progress only shows 89% is strange, do you think you can share the data you used as well?

**Comment by shrit** [comment]
@zoq I can try to send you a part of the data by email.
I have decreased the tolerance to -1 as you have requested. 
Is it ok to send you the data by the email shown on your profile page?

I have changed Adam to SGD, which uses Adam update rule, The problem still here even after using SGD and it stops also on `89%`. However, the loss decreased dramatically using SGD which I do not understand
```
 //  ens::AdamType<ens::AdamUpdate> optimizer;
  ens::SGD<ens::AdamUpdate> optimizer;
```

**Comment by shrit** [comment]
@zoq  You are right, something has to do with the dataset (which is also strange).
For example, I have changed the data set with a much smaller one.
In this case, the epoch stops at `25/794` and complete to `100%`
It still strange, because the loss continues to decrease when the second `for loop (i=1, i=2, ...)`

Please, have a look at the training after the epoche `25`, it stops at `98%`
```
1.70999 [====================================================================================================] 100% - ETA: 0s - loss: 1.6839
126/126 [====================================================================================================] 100% - 0s 6ms/step - loss: 1.70999
Epoch 22/794
1.57015 [====================================================================================================] 100% - ETA: 0s - loss: 1.54619
126/126 [====================================================================================================] 100% - 0s 6ms/step - loss: 1.57015
Epoch 23/794
1.52725 [====================================================================================================] 100% - ETA: 0s - loss: 1.50395
126/126 [====================================================================================================] 100% - 1s 8ms/step - loss: 1.52725
Epoch 24/794
1.44361 [====================================================================================================] 100% - ETA: 0s - loss: 1.42159
126/126 [====================================================================================================] 100% - 0s 7ms/step - loss: 1.44361
Epoch 25/794
Train function value:  5391.46=============================================================================>.] 98% - ETA: 0s - loss: 1.34786
Evaluate function value : 1.22672e-29
Epoch 1/794
1.40014 [====================================================================================================] 100% - ETA: 0s - loss: 1.37878
126/126 [====================================================================================================] 100% - 0s 6ms/step - loss: 1.40014
Epoch 2/794
1.29211 [====================================================================================================] 100% - ETA: 0s - loss: 1.27239
126/126 [====================================================================================================] 100% - 0s 7ms/step - loss: 1.29211
Epoch 3/794
1.22548 [====================================================================================================] 100% - ETA: 0s - loss: 1.20678
126/126 [====================================================================================================] 100% - 0s 7ms/step - loss: 1.22548
Epoch 4/794
1.23075 [====================================================================================================] 100% - ETA: 0s - loss: 1.21198
126/126 [====================================================================================================] 100% - 0s 7ms/step - loss: 1.23075
Epoch 5/794
1.16885 [====================================================================================================] 100% - ETA: 0s - loss: 1.15102
126/126 [====================================================================================================] 100% - 0s 7ms/step - loss: 1.16885
Epoch 6/794
1.11097 [====================================================================================================] 100% - ETA: 0s - loss: 1.09403
126/126 [====================================================================================================] 100% - 0s 6ms/step - loss: 1.11097

```
I have measured the execution time for both dataset, it seems that the train function return after executing for around 24 seconds, I do not think there is any limit or timer Right?.

**Comment by zoq** [comment]
You are right, there is no time limit, what is the size of the dataset?

**Comment by shrit** [comment]
About 12 mega bytes

**Comment by zoq** [comment]
Sorry for the confusion, I was talking about number of rows, cols.

**Comment by shrit** [comment]
Ok, consider row major matrix, 120k rows, and 23 columns

**Comment by zoq** [comment]
Okay, thanks, let's see if I can reproduce the issue, just by using some dummy dataset that has the same dimensions.

**Comment by shrit** [comment]
@zoq Do you have any new results? Can I help?

**Comment by zoq** [comment]
Sorry, I have forgotten to look into this, unfortunately I'm not sure I can look into this today, but definitely tomorrow. Also, can you send me a subset of the data per mail?

**Comment by shrit** [comment]
Ok I will send you the dataset

**Comment by zoq** [comment]
Thanks, so I haven't seen that you haven't configured the number of iterations: `maxIterations` which is per default set to 100000 which is less than the number of samples in your dataset. So you could either increase the number of iterations or set it to 0 which will not limit the number of iterations. Let me know if I should clarify anything.

**Comment by shrit** [comment]
@zoq, That resolve part of the problem.
However, why the epoch number does not increase?

**Comment by shrit** [comment]
@zoq, I think I understand better now. 
Actually the `epoch` number is related to the `MaxIterations` number.
`epoch = MaxIterations / dataset.size()`
If this the case, What does the max epoch number represent if it can not be reached automatically?
For example, in my case, when `MaxIterations` was equal to `100000`, the max epoch was 794. After I increased the MaxIterations to 1000000 the max epoch now is 7937.
I think the training will stop at around epoch=250, in this case what does the Max epoch represent?

**Comment by zoq** [comment]
> @zoq, I think I understand better now.
> Actually the epoch number is related to the MaxIterations number.
> epoch = MaxIterations / dataset.size()
> If this the case, What does the max epoch number represent if it can not be reached automatically?
> 

That depends on the number of samples in the dataset, it's basically a default value, which has to be tailored down in each case. But I can see the confusion. Do you think it would be helpful to use max(MaxIterations, dataset.size()) as the default value,? In which case we make sure we iterate at least once over the entire dataset? @rcurtin what do you think?

> I can understand that epochs are related to the epoch size which is related to NumFunctions and BatchSize is there any recommended way to regulate them in order to make epoch reach epochs?

You mean, not returning before we reached the number of epochs defined?

**Comment by shrit** [comment]
> That depends on the number of samples in the dataset, it's basically a default value, which has to be tailored down in each case. But I can see the confusion. Do you think it would be helpful to use max(MaxIterations, dataset.size()) as the default value,? In which case we make sure we iterate at least once over the entire dataset? @rcurtin what do you think?

I agree with you,  at least we need to iterate once thought out the entire dataset.

> > I can understand that epochs are related to the epoch size which is related to NumFunctions and BatchSize is there any recommended way to regulate them in order to make epoch reach epochs?
> 
> You mean, not returning before we reached the number of epochs defined?

Exactly. Why?
Otherwise the number of epochs defined does not mean a lot of things am I right?

**Comment by rcurtin** [comment]
> That depends on the number of samples in the dataset, it's basically a default value, which has to be tailored down in each case. But I can see the confusion. Do you think it would be helpful to use max(MaxIterations, dataset.size()) as the default value,? In which case we make sure we iterate at least once over the entire dataset? @rcurtin what do you think?

This one is a little bit tricky---most neural network literature and software uses the idea of an "epoch", or a full pass over the dataset, but the more general optimization literature doesn't necessarily do that.  So personally I think the idea of an "epoch" is a little bit specific to deep learning and machine learning, but ensmallen isn't necessarily specific to those fields.

One solution might be to provide both: the optimizers could have parameters `maxIterations` and `epochs`, and if `epochs` is specified, then `maxIterations` is ignored.  I'm not sure exactly how that might work, but it's one idea.  What do you think?

**Comment by shrit** [comment]
@zoq keep it open?

**Comment by zoq** [comment]
Good idea, let's keep this open.

**Comment by rcurtin** [comment]
It seems like we did not decide on what a solution should be here or what we should do in order to close the issue.  Maybe we should start by defining what we need to change or improve?

**Comment by shrit** [comment]
I think there are two things to consider from this issue:
1) Be sure to iterate and achieve a full pass on the entire dataset.
2) Understand why `Train()`is  returning before we reached the number of epochs defined

**Comment by rcurtin** [comment]
Thanks for the comments; actually it looks like when things are terminating early, it is because the convergence criterion is met.  One thing that you could do from your end to prevent that is set `tolerance` to `-1`, or something like that, so that the optimizer will never terminate due to a convergence check.  I don't think there's any need to change the behavior of the optimizers in ensmallen for that though.

For (2), I think we can make an improvement.  I opened https://github.com/mlpack/ensmallen/issues/150 to handle that.  If you agree that that's a good solution, we can close this issue here and handle the rest of it there. :+1:

**Comment by shrit** [comment]
Thanks for your comments. 
I think it is better to keep this issue open as long as https://github.com/mlpack/ensmallen/issues/150 still open.
For (2) you are right, maybe the optimizer is reaching convergence. However, when I set `tolerance` to `-1` I have noticed on the long run that the optimizer might converge even better,
For (1) When I set maxIteration to the size of the dataset, it will read the entire dataset. Especially,  when the dataset size is higher than the default value which is `100000`.
I think that `maxIteration` need to be set when we parse the number of columns of the dataset

**Comment by rcurtin** [comment]
Sure, no problem, we can keep this open until ensmallen#150 is resolved.

It's true that different convergence strategies can produce different results, and sometimes it can be better to run to the limit of iterations.  But that's totally dependent on the problem.

For (1), I can see why from a neural network perspective one would definitely want at least one scan over the dataset.  But that doesn't always apply for non-neural-network problems.  So finding something that works in both situations is a bit tricky.  Do you think it would be sufficient if a warning was printed by ensmallen when `Optimize()` was called but `maxIterations` was set to be less than the number of points in the dataset?  Alternately, we could put this warning in mlpack's ANN code itself; it would be easy to have mlpack detect when the optimizer's maximum number of iterations was less than the number of points in the dataset.

**Comment by shrit** [comment]
Actually, it is a good idea to print the warning in the ANN code itself.
For now, the user can have an indication whether the neural network is reading the dataset.
Otherwise, we can fix the `maxIteration` in the ANN code when the user enters the training dataset in the training function.

**Comment by zoq** [comment]
Agreed, printing some warning in the ANN code would be a clean 'solution'. I'll open an issue for the idea later.

**Comment by rcurtin** [comment]
@zoq just checking in, did you have a chance to open an issue?  I don't see anything referenced here; just wanted to make sure it happened. :)  If you haven't had a chance I'm happy to do it. :+1:

**Comment by shrit** [comment]
@rcurtin @zoq I will open an issue for that in ensmallen. I will also try to make a pull request for it if I had time in the next few days.

**Comment by shrit** [comment]
I think this one can be closed now
